# 🧠 Text Classification: Fiction vs. Non-Fiction

In this section, we enhance our recommendation system by adding **genre-level classification** — specifically distinguishing between **Fiction** and **Non-Fiction** books.

From our earlier data exploration, we observed that the dataset contains over **500 unique categories**, many of which are **extremely niche** and associated with only a single book. This makes traditional multi-class classification difficult and noisy.

To address this, we simplify the task by using the **book description** to predict whether a title belongs to **Fiction** or **Non-Fiction**. This not only improves the clarity of our dataset but also enables more effective filtering and personalization.

### 🧪 Zero-Shot Classification

We utilize **zero-shot text classification** from Hugging Face — specifically, models like `facebook/bart-large-mnli` — to perform this task without needing to fine-tune or train on labeled data.

This method allows us to:
- Classify any book based on its description
- Avoid manual labeling
- Use **pre-trained large language models (LLMs)** to generalize across diverse topics

This simple yet powerful classification will help narrow down recommendations and enhance the overall search experience.


In [1]:
!pip install transformers

In [2]:
# get the HuggingFace API key from Google Secrets Manager
from google.colab import userdata
HUGGINGFACEHUB_API_TOKEN = userdata.get('HUGGINGFACEHUB_API_TOKEN')

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
import pandas as pd

# Replace 'path/to/your/book_cleaned.csv' with the actual path to your file in Google Drive
file_path = '/content/drive/MyDrive/Colab Notebooks/Semantic Book Recommender/books_cleaned.csv'

try:
    books = pd.read_csv(file_path)
    print("CSV file loaded successfully!")
    display(books.head())
except FileNotFoundError:
    print(f"Error: The file was not found at {file_path}")
except Exception as e:
    print(f"An error occurred: {e}")

CSV file loaded successfully!


isbn13      isbn10                title  \
0  9780002005883  0002005883               Gilead   
1  9780002261982  0002261987         Spider's Web   
2  9780006178736  0006178731       Rage of angels   
3  9780006280897  0006280897       The Four Loves   
4  9780006280934  0006280935  The Problem of Pain   

                           authors                     categories  \
0               Marilynne Robinson                        Fiction   
1  Charles Osborne;Agatha Christie  Detective and mystery stories   
2                   Sidney Sheldon                        Fiction   
3              Clive Staples Lewis                 Christian life   
4              Clive Staples Lewis                 Christian life   

                                           thumbnail  \
0  http://books.google.com/books/content?id=KQZCP...   
1  http://books.google.com/books/content?id=gA5GP...   
2  http://books.google.com/books/content?id=FKo2T...   
3  http://books.google.com/books/content?id=XhQ5X...   
4  http://books.google.com/books/content?id=Kk-uV...   

                                         description  published_year  \
0  A NOVEL THAT READERS and critics have been eag...          2004.0   
1  A new 'Christie for Christmas' -- a full-lengt...          2000.0   
2  A memorable, mesmerizing heroine Jennifer -- b...          1993.0   
3  Lewis' work on the nature of love divides love...          2002.0   
4  "In The Problem of Pain, C.S. Lewis, one of th...          2002.0   

   average_rating  num_pages  ratings_count     title_and_subtiles  \
0            3.85      247.0          361.0                 Gilead   
1            3.83      241.0         5164.0  Spider's Web: A Novel   
2            3.93      512.0        29532.0         Rage of angels   
3            4.15      170.0        33684.0         The Four Loves   
4            4.09      176.0        37569.0    The Problem of Pain   

                                  tagged_description  
0  9780002005883 A NOVEL THAT READERS and critics...  
1  9780002261982 A new 'Christie for Christmas' -...  
2  9780006178736 A memorable, mesmerizing heroine...  
3  9780006280897 Lewis' work on the nature of lov...  
4  9780006280934 "In The Problem of Pain, C.S. Le...

In [5]:
# Count the occurrences of each book category
category_counts = books['categories'].value_counts()

# Display the category counts
print("Book Category Counts:")
display(category_counts)

Book Category Counts:


categories
Fiction                       2111
Juvenile Fiction               390
Biography & Autobiography      311
History                        207
Literary Criticism             124
                              ... 
Human-animal relationships       1
Imperialism                      1
Aged women                       1
Humorous stories                 1
Butlers                          1
Name: count, Length: 479, dtype: int64

Analysis on the books categories, and how there's are way more fictions and very niche books.

In [6]:
# Count the occurrences of each book category
category_counts = books['categories'].value_counts()

# Filter categories with 50 or more books
filtered_categories = category_counts[category_counts >= 50]

# Display the filtered category counts
print("Book Categories with 50 or More Books:")
display(filtered_categories)

Book Categories with 50 or More Books:


categories
Fiction                      2111
Juvenile Fiction              390
Biography & Autobiography     311
History                       207
Literary Criticism            124
Religion                      117
Philosophy                    117
Comics & Graphic Novels       116
Drama                          86
Juvenile Nonfiction            57
Science                        56
Poetry                         51
Literary Collections           50
Name: count, dtype: int64

In [7]:
books[books["categories"] == "Juvenile Fiction"]

isbn13      isbn10  \
30    9780006646006  000664600X   
79    9780020442608  0020442602   
85    9780030547744  0030547741   
86    9780060000141  0060000147   
87    9780060001537  0060001534   
...             ...         ...   
4845  9781590385814  1590385810   
4930  9781596792500  1596792507   
4942  9781599900056  159990005X   
5010  9781844580514  1844580512   
5157  9781932416374  1932416374   

                                                  title  \
30                                   Ocean Star Express   
79                       The voyage of the Dawn Treader   
85                             Where the Red Fern Grows   
86                                       Poppy's Return   
87                                    Diary of a Spider   
...                                                 ...   
4845                                         Fablehaven   
4930  Sherlock Holmes and the Case of the Hound of t...   
4942                                    The Drift House   
5010                               Attack of the Jaguar   
5157                The Very Persistent Gappers of Frip   

                                  authors        categories  \
30               Mark Haddon;Peter Sutton  Juvenile Fiction   
79                    Clive Staples Lewis  Juvenile Fiction   
85                           Wilson Rawls  Juvenile Fiction   
86                                    Avi  Juvenile Fiction   
87                          Doreen Cronin  Juvenile Fiction   
...                                   ...               ...   
4845                         Brandon Mull  Juvenile Fiction   
4930  Arthur Conan Doyle;Malvina G. Vogel  Juvenile Fiction   
4942                            Dale Peck  Juvenile Fiction   
5010                         M. A. Harvey  Juvenile Fiction   
5157           George Saunders;Lane Smith  Juvenile Fiction   

                                              thumbnail  \
30    http://books.google.com/books/content?id=I2QZA...   
79    http://books.google.com/books/content?id=fDD3C...   
85    http://books.google.com/books/content?id=IHpRw...   
86    http://books.google.com/books/content?id=XbcMJ...   
87    http://books.google.com/books/content?id=UWvZo...   
...                                                 ...   
4845  http://books.google.com/books/content?id=tbVIP...   
4930  http://books.google.com/books/content?id=EWgWP...   
4942  http://books.google.com/books/content?id=kbwPY...   
5010  http://books.google.com/books/content?id=3HUdt...   
5157  http://books.google.com/books/content?id=xYJOP...   

                                            description  published_year  \
30    Joe and his parents are enjoying a summer holi...          2002.0   
79    The "Dawn Treader" is the first ship Narnia ha...          1970.0   
85    A young boy living in the Ozarks achieves his ...          2000.0   
86    There's trouble at Gray House, the girlhood ho...          2006.0   
87    This is the diary ... of a spider. But don't b...          2005.0   
...                                                 ...             ...   
4845  When Kendra and Seth go to stay at their grand...          2006.0   
4930  Sherlock Holmes and Dr. Watson travel to the b...          2005.0   
4942  Sent to stay with their uncle in a ship-like h...          2006.0   
5010  This training manual for operatives of Xtreme ...          2004.0   
5157  A novel set in the seaside village of Frip fol...          2006.0   

      average_rating  num_pages  ratings_count  \
30              3.50       32.0            1.0   
79              4.09      216.0         2869.0   
85              4.37      288.0           95.0   
86              3.99      256.0         1086.0   
87              4.25       40.0         7903.0   
...              ...        ...            ...   
4845            4.09      351.0       111896.0   
4930            4.51      237.0           28.0   
4942            3.64      437.0          595.0   
5010            

In [8]:
books[books["categories"] == "Juvenile Nonfiction"]

isbn13      isbn10  \
107   9780060277406  0060277408   
108   9780060278427  0060278420   
228   9780060782139  0060782137   
267   9780060882600  0060882603   
434   9780064462044  0064462048   
435   9780064462341  006446234X   
439   9780066236179  0066236177   
812   9780142302279  0142302279   
821   9780142407226  0142407224   
823   9780142407929  0142407925   
1014  9780198320272  0198320272   
1074  9780237525378  0237525372   
1739  9780375812934  0375812938   
1742  9780375823787  0375823786   
1770  9780380732548  0380732548   
1972  9780395776087  0395776082   
2105  9780439314558  0439314550   
2117  9780439420068  0439420067   
2139  9780439736466  0439736463   
2183  9780440229353  0440229359   
2211  9780440419174  0440419174   
2320  9780448437439  0448437430   
2323  9780448444826  0448444828   
2503  9780486276809  0486276805   
2509  9780486296388  0486296385   
2609  9780521618748  0521618746   
2615  9780525469780  0525469788   
2858  9780590032490  0590032496   
2859  9780590103879  0590103873   
2897  9780618111206  0618111204   
3133  9780679890522  0679890521   
3202  9780689833762  0689833768   
3203  9780689834196  0689834195   
3204  9780689835612  0689835612   
3219  9780689869907  0689869908   
3223  9780689876707  068987670X   
3263  9780723247791  072324779X   
3270  9780736805995  0736805990   
3292  9780739844328  0739844326   
3522  9780753459645  0753459647   
3552  9780756614195  0756614198   
3565  9780761417101  0761417109   
3797  9780789458209  0789458209   
3798  9780791070482  0791070484   
3799  9780791081266  0791081265   
3800  9780792241737  0792241738   
3861  9780805069853  0805069852   
3863  9780805071061  0805071067   
3921  9780810935204  0810935201   
4069  9780822560562  0822560569   
4072  9780823410781  0823410781   
4409  9781400307845  1400307848   
4596  9781560067849  1560067845   
4744  9781580495868  1580495869   
4808  9781586853334  1586853333   
5041  9781857232097  1857232097   
5142  9781931414166  1931414165   

                                                  title  \
107                          The Secret Garden Cookbook   
108                                       Laura's Album   
228                         Time For Kids: Butterflies!   
267                       The Annotated Charlotte's Web   
434                         My Little House Crafts Book   
435                                        Pioneer Girl   
439                    A Light in the Attic Book and CD   
812                                        Dirty Beasts   
821                      The Tough Guide to Fantasyland   
823                                             The BFG   
1014                                      Julius Caesar   
1074                                       Oliver Twist   
1739       Jurassic Park Institute Dinosaur Field Guide   
1742                    Ancient Greece and the Olympics   
1770                                         Anne Frank   
1972                                 An American Plague   
2105                    Conversations with J.K. Rowling   
2117                                               Eat!   
2139                          Where We Are, what We See   
2183                    How Angel Peterson Got His Name   
2211                                   Vincent Van Gogh   
2320                           Who Was John F. Kennedy?   
2323                                Who Was Anne Frank?   
2503                    The Secret Garden Coloring Book   
2509                         Ballet Class Coloring Book   
2609                                             Hamlet   
2615                Boy, Were We Wrong about Dinosaurs!   
2858                                        The witches   
2859                                 Expository Writing   
2897                                        Once a Wolf   
3133                                 The Trail of Tears   
3202              Sad Underwear and Other Complications   
3203                               

### 📚 Grouping Categories for Better Classification

While the dataset contains many niche and granular categories, some of them can still be useful — especially if we **group related labels into broader, more practical categories**.

For example:
- `'Biography & Autobiography'` can clearly be mapped to **"Non-Fiction"**
- Categories like `'Fantasy'`, `'Historical Fiction'`, or `'Mystery & Detective'` can be grouped under **"Fiction"**

By applying this kind of **high-level grouping**, we can:
- Reduce the number of distinct categories
- Simplify the classification task
- Improve model interpretability and performance

This grouped mapping can serve as ground truth for validating our **zero-shot classification**, or as a **fallback filter** when no description is available.



In [9]:
category_mapping = {
    'Fiction': "Fiction",
    'Juvenile Fiction': "Children's Fiction",
    'Biography & Autobiography': "Nonfiction",
    'History': "Nonfiction",
    'Literary Criticism': "Nonfiction",
    'Philosophy': "Nonfiction",
    'Religion': "Nonfiction",
    'Comics & Graphic Novels': "Fiction",
    'Drama': "Fiction",
    'Juvenile Nonfiction': "Children's Nonfiction",
    'Science': "Nonfiction",
    'Poetry': "Fiction"
}

books["simple_categories"] = books["categories"].map(category_mapping)

In [10]:
books

isbn13      isbn10  \
0     9780002005883  0002005883   
1     9780002261982  0002261987   
2     9780006178736  0006178731   
3     9780006280897  0006280897   
4     9780006280934  0006280935   
...             ...         ...   
5192  9788172235222  8172235224   
5193  9788173031014  8173031010   
5194  9788179921623  817992162X   
5195  9788185300535  8185300534   
5196  9789027712059  9027712050   

                                                  title  \
0                                                Gilead   
1                                          Spider's Web   
2                                        Rage of angels   
3                                        The Four Loves   
4                                   The Problem of Pain   
...                                                 ...   
5192                                  Mistaken Identity   
5193                                Journey to the East   
5194  The Monk Who Sold His Ferrari: A Fable About F...   
5195                                          I Am that   
5196                           The Berlin Phenomenology   

                                           authors  \
0                               Marilynne Robinson   
1                  Charles Osborne;Agatha Christie   
2                                   Sidney Sheldon   
3                              Clive Staples Lewis   
4                              Clive Staples Lewis   
...                                            ...   
5192                              Nayantara Sahgal   
5193                                 Hermann Hesse   
5194                                  Robin Sharma   
5195  Sri Nisargadatta Maharaj;Sudhakar S. Dikshit   
5196                 Georg Wilhelm Friedrich Hegel   

                         categories  \
0                           Fiction   
1     Detective and mystery stories   
2                           Fiction   
3                    Christian life   
4                    Christian life   
...                             ...   
5192        Indic fiction (English)   
5193              Adventure stories   
5194               Health & Fitness   
5195                     Philosophy   
5196                        History   

                                              thumbnail  \
0     http://books.google.com/books/content?id=KQZCP...   
1     http://books.google.com/books/content?id=gA5GP...   
2     http://books.google.com/books/content?id=FKo2T...   
3     http://books.google.com/books/content?id=XhQ5X...   
4     http://books.google.com/books/content?id=Kk-uV...   
...                                                 ...   
5192  http://books.google.com/books/content?id=q-tKP...   
5193  http://books.google.com/books/content?id=rq6JP...   
5194  http://books.google.com/books/content?id=c_7mf...   
5195  http://books.google.com/books/content?id=Fv_JP...   
5196  http://books.google.com/books/content?id=Vy7Sk...   

                                            description  published_year  \
0     A NOVEL THAT READERS and critics have been eag...          2004.0   
1     A new 'Christie for Christmas' -- a full-lengt...          2000.0   
2     A memorable, mesmerizing heroine Jennifer -- b...          1993.0   
3     Lewis' work on the nature of love divides love...          2002.0   
4     "In The Problem of Pain, C.S. Lewis, one of th...          2002.0   
...                                                 ...             ...   
5192  On A Train Journey Home To North India After L...          2003.0   
5193  This book tells the tale of a man who goes on ...          2002.0   
5194  Wisdom to Create a Life of Passion, Purpose, a...          2003.0   
5195  This collection of the timeless teachings of o...          1999.0   
5196  Since the three volume edition ofHegel's Philo...          1981.0   

      average_rating  num_pages  ratings_count  \
0               3.85      247.0          361.0   
1               3.83      241.0         5164.0   
2    

In [11]:
books[~(books["simple_categories"].isna())]

isbn13      isbn10                      title  \
0     9780002005883  0002005883                     Gilead   
2     9780006178736  0006178731             Rage of angels   
8     9780006482079  0006482074        Warhost of Vastmark   
30    9780006646006  000664600X         Ocean Star Express   
46    9780007121014  0007121016         Taken at the Flood   
...             ...         ...                        ...   
5178  9781933648279  1933648279  Night Has a Thousand Eyes   
5188  9784770028969  4770028962         Coin Locker Babies   
5189  9788122200850  8122200850           Cry, the Peacock   
5195  9788185300535  8185300534                  I Am that   
5196  9789027712059  9027712050   The Berlin Phenomenology   

                                           authors        categories  \
0                               Marilynne Robinson           Fiction   
2                                   Sidney Sheldon           Fiction   
8                                      Janny Wurts           Fiction   
30                        Mark Haddon;Peter Sutton  Juvenile Fiction   
46                                 Agatha Christie           Fiction   
...                                            ...               ...   
5178                              Cornell Woolrich           Fiction   
5188                                           村上龍           Fiction   
5189                                   Anita Desai           Fiction   
5195  Sri Nisargadatta Maharaj;Sudhakar S. Dikshit        Philosophy   
5196                 Georg Wilhelm Friedrich Hegel           History   

                                              thumbnail  \
0     http://books.google.com/books/content?id=KQZCP...   
2     http://books.google.com/books/content?id=FKo2T...   
8     http://books.google.com/books/content?id=uOL0f...   
30    http://books.google.com/books/content?id=I2QZA...   
46    http://books.google.com/books/content?id=3gWlx...   
...                                                 ...   
5178  http://books.google.com/books/content?id=3Gk6s...   
5188  http://books.google.com/books/content?id=87DJw...   
5189  http://books.google.com/books/content?id=_QKwV...   
5195  http://books.google.com/books/content?id=Fv_JP...   
5196  http://books.google.com/books/content?id=Vy7Sk...   

                                            description  published_year  \
0     A NOVEL THAT READERS and critics have been eag...          2004.0   
2     A memorable, mesmerizing heroine Jennifer -- b...          1993.0   
8     Tricked once more by his wily half-brother, Ly...          1995.0   
30    Joe and his parents are enjoying a summer holi...          2002.0   
46    A Few Weeks After Marrying An Attractive Young...          2002.0   
...                                                 ...             ...   
5178  "Cornell Woolrich's novels define the essence ...          2007.0   
5188  Rescued from the lockers in which they were le...          2002.0   
5189  This book is the story of a young girl obsesse...          1980.0   
5195  This collection of the timeless teachings of o...          1999.0   
5196  Since the three volume edition ofHegel's Philo...          1981.0   

      average_rating  num_pages  ratings_count  \
0               3.85      247.0          361.0   
2               3.93      512.0        29532.0   
8               4.03      522.0         2966.0   
30              3.50       32.0            1.0   
46              3.71      352.0         8852.0   
...              ...        ...            ...   
5178            3.77      344.0          680.0   
5188            3.75      393.0         5560.0   
5189            3.22      218.0          134.0   
5195            4.51      531.0          104.0   
5196            0.00      210.0            0.0   

                                  title_and_subtiles  \
0                                             Gilead   
2                                     Rage of angels   
8                       

We will be using **Hugging Face Transformers** for our **zero-shot text classification** task.

This allows us to leverage powerful pre-trained language models — such as `facebook/bart-large-mnli` — to classify book descriptions into categories like **Fiction** and **Non-Fiction**, without requiring any additional labeled training data.


In [12]:
from transformers import pipeline
import torch

device = 0 if torch.cuda.is_available() else -1

classifier = pipeline("zero-shot-classification", model="facebook/bart-large-mnli", device=device)

if device == 0:
    print("GPU is available and will be used for the pipeline.")
else:
    print("GPU is not available, the pipeline will run on CPU.")

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0


GPU is available and will be used for the pipeline.


In [13]:
fiction_categories = ["Fiction", "Nonfiction"]

In [14]:
sequence = books.loc[books["simple_categories"] == "Fiction", "description"].reset_index(drop=True)[0]
sequence

'A NOVEL THAT READERS and critics have been eagerly anticipating for over a decade, Gilead is an astonishingly imagined story of remarkable lives. John Ames is a preacher, the son of a preacher and the grandson (both maternal and paternal) of preachers. It’s 1956 in Gilead, Iowa, towards the end of the Reverend Ames’s life, and he is absorbed in recording his family’s story, a legacy for the young son he will never see grow up. Haunted by his grandfather’s presence, John tells of the rift between his grandfather and his father: the elder, an angry visionary who fought for the abolitionist cause, and his son, an ardent pacifist. He is troubled, too, by his prodigal namesake, Jack (John Ames) Boughton, his best friend’s lost son who returns to Gilead searching for forgiveness and redemption. Told in John Ames’s joyous, rambling voice that finds beauty, humour and truth in the smallest of life’s details, Gilead is a song of celebration and acceptance of the best and the worst the world ha

In [15]:
classifier(sequence, fiction_categories)

{'sequence': 'A NOVEL THAT READERS and critics have been eagerly anticipating for over a decade, Gilead is an astonishingly imagined story of remarkable lives. John Ames is a preacher, the son of a preacher and the grandson (both maternal and paternal) of preachers. It’s 1956 in Gilead, Iowa, towards the end of the Reverend Ames’s life, and he is absorbed in recording his family’s story, a legacy for the young son he will never see grow up. Haunted by his grandfather’s presence, John tells of the rift between his grandfather and his father: the elder, an angry visionary who fought for the abolitionist cause, and his son, an ardent pacifist. He is troubled, too, by his prodigal namesake, Jack (John Ames) Boughton, his best friend’s lost son who returns to Gilead searching for forgiveness and redemption. Told in John Ames’s joyous, rambling voice that finds beauty, humour and truth in the smallest of life’s details, Gilead is a song of celebration and acceptance of the best and the worst

In [16]:
import numpy as np

# get the max of the scores from the classsifier
max_index = np.argmax(classifier(sequence, fiction_categories)["scores"])
max_label = classifier(sequence, fiction_categories)["labels"][max_index]

max_label

'Fiction'

In [17]:
# make a re-useable function
def generate_prediction(sequence, categories):
    prediction = classifier(sequence, categories)
    max_index = np.argmax(classifier(sequence, fiction_categories)["scores"])
    max_label = classifier(sequence, fiction_categories)["labels"][max_index]
    return max_label

### ✅ Evaluating Model Performance: Predicted vs. Actual Categories

To determine whether our **zero-shot classification model** is reliable, we will compare the **predicted labels** (Fiction vs. Non-Fiction) with the **actual genre information** from the dataset — after mapping detailed genres to high-level categories.

This evaluation will help us understand:
- How often the model aligns with the known categories
- Whether the predictions are **consistent enough** to use in production
- If further refinement (e.g., rule-based correction or post-filtering) is needed


In [18]:
from tqdm import tqdm

actual_cats = []
predicted_cats = []

for i in tqdm(range(0, 300)):
  sequence = books.loc[books["simple_categories"] == "Fiction", "description"].reset_index(drop=True)[i]
  predicted_cats += [generate_prediction(sequence, fiction_categories)]
  actual_cats += ["Fiction"]


100%|██████████| 300/300 [01:16<00:00,  3.91it/s]


In [19]:
# now we do the same for non fiction
for i in tqdm(range(0, 300)):
  sequence = books.loc[books["simple_categories"] == "Nonfiction", "description"].reset_index(drop=True)[i]
  predicted_cats += [generate_prediction(sequence, fiction_categories)]
  actual_cats += ["Nonfiction"]

100%|██████████| 300/300 [01:20<00:00,  3.72it/s]


In [20]:
predictions_df = pd.DataFrame({"actual": actual_cats, "predicted": predicted_cats})

In [21]:
predictions_df.head()

actual   predicted
0  Fiction     Fiction
1  Fiction     Fiction
2  Fiction     Fiction
3  Fiction  Nonfiction
4  Fiction     Fiction

In [22]:
predictions_df["correct_prediction"] = (
    np.where(predictions_df["actual"] == predictions_df["predicted"], 1, 0)
)

In [23]:
predictions_df["correct_prediction"].sum() / len(predictions_df)

np.float64(0.7783333333333333)

As we can see, the model is able to predict around **77.8%** correctly. This is a **decent model** and a **good result for a pre-built LLM**. However, **1 in 5 predictions is still incorrect**, which shows that while the model is useful, it’s not perfect.


In [24]:
isbns = []
predicted_cats = []

missing_cats = books.loc[books["simple_categories"].isna(), ["isbn13", "description"]].reset_index(drop=True)

In [25]:
isbns = []
predicted_cats = []

for i in tqdm(range(0, len(missing_cats))):
  sequence = missing_cats["description"][i]
  try:
    predicted_cat = generate_prediction(sequence, fiction_categories)
    isbns.append(missing_cats["isbn13"][i])
    predicted_cats.append(predicted_cat)
  except Exception as e:
    print(f"Error processing sequence {i}: {e}")
    # If an error occurs, we skip appending to both lists for this iteration

100%|██████████| 1454/1454 [06:07<00:00,  3.95it/s]


In [26]:
print(f"Length of isbns: {len(isbns)}")
print(f"Length of predicted_cats: {len(predicted_cats)}")

Length of isbns: 1454
Length of predicted_cats: 1454


In [27]:
missing_predictions_df = pd.DataFrame({"isbn": isbns, "predicted": predicted_cats})

In [28]:
missing_predictions_df

isbn   predicted
0     9780002261982     Fiction
1     9780006280897  Nonfiction
2     9780006280934  Nonfiction
3     9780006380832  Nonfiction
4     9780006470229     Fiction
...             ...         ...
1449  9788125026600  Nonfiction
1450  9788171565641     Fiction
1451  9788172235222     Fiction
1452  9788173031014  Nonfiction
1453  9788179921623     Fiction

[1454 rows x 2 columns]

Now that we have some confidence in the zero-shot model, we will use it to **fill in missing categories** based on each book’s description.

Previously, if a book did not fall into a recognizable or simple category (e.g., Fiction or Non-Fiction), it was marked as **NA**. Now, we will use the **predicted label from the model** to populate these missing values. This will help enrich the dataset and improve filtering in later stages.


In [29]:
books = pd.merge(books, missing_predictions_df, left_on="isbn13", right_on="isbn", how="left")

In [30]:
books

isbn13      isbn10  \
0     9780002005883  0002005883   
1     9780002261982  0002261987   
2     9780006178736  0006178731   
3     9780006280897  0006280897   
4     9780006280934  0006280935   
...             ...         ...   
5192  9788172235222  8172235224   
5193  9788173031014  8173031010   
5194  9788179921623  817992162X   
5195  9788185300535  8185300534   
5196  9789027712059  9027712050   

                                                  title  \
0                                                Gilead   
1                                          Spider's Web   
2                                        Rage of angels   
3                                        The Four Loves   
4                                   The Problem of Pain   
...                                                 ...   
5192                                  Mistaken Identity   
5193                                Journey to the East   
5194  The Monk Who Sold His Ferrari: A Fable About F...   
5195                                          I Am that   
5196                           The Berlin Phenomenology   

                                           authors  \
0                               Marilynne Robinson   
1                  Charles Osborne;Agatha Christie   
2                                   Sidney Sheldon   
3                              Clive Staples Lewis   
4                              Clive Staples Lewis   
...                                            ...   
5192                              Nayantara Sahgal   
5193                                 Hermann Hesse   
5194                                  Robin Sharma   
5195  Sri Nisargadatta Maharaj;Sudhakar S. Dikshit   
5196                 Georg Wilhelm Friedrich Hegel   

                         categories  \
0                           Fiction   
1     Detective and mystery stories   
2                           Fiction   
3                    Christian life   
4                    Christian life   
...                             ...   
5192        Indic fiction (English)   
5193              Adventure stories   
5194               Health & Fitness   
5195                     Philosophy   
5196                        History   

                                              thumbnail  \
0     http://books.google.com/books/content?id=KQZCP...   
1     http://books.google.com/books/content?id=gA5GP...   
2     http://books.google.com/books/content?id=FKo2T...   
3     http://books.google.com/books/content?id=XhQ5X...   
4     http://books.google.com/books/content?id=Kk-uV...   
...                                                 ...   
5192  http://books.google.com/books/content?id=q-tKP...   
5193  http://books.google.com/books/content?id=rq6JP...   
5194  http://books.google.com/books/content?id=c_7mf...   
5195  http://books.google.com/books/content?id=Fv_JP...   
5196  http://books.google.com/books/content?id=Vy7Sk...   

                                            description  published_year  \
0     A NOVEL THAT READERS and critics have been eag...          2004.0   
1     A new 'Christie for Christmas' -- a full-lengt...          2000.0   
2     A memorable, mesmerizing heroine Jennifer -- b...          1993.0   
3     Lewis' work on the nature of love divides love...          2002.0   
4     "In The Problem of Pain, C.S. Lewis, one of th...          2002.0   
...                                                 ...             ...   
5192  On A Train Journey Home To North India After L...          2003.0   
5193  This book tells the tale of a man who goes on ...          2002.0   
5194  Wisdom to Create a Life of Passion, Purpose, a...          2003.0   
5195  This collection of the timeless teachings of o...          1999.0   
5196  Since the three volume edition ofHegel's Philo...          1981.0   

      average_rating  num_pages  ratings_count  \
0               3.85      247.0          361.0   
1               3.83      241.0         5164.0   
2    

In [31]:
# replace the simple categories with the predicted or original simple categories
books["simple_categories"] = np.where(books["simple_categories"].isna(), books["predicted"], books["simple_categories"])
# we don't need the predicted categories anymore
books = books.drop(columns = ["predicted"])

books.head()

isbn13      isbn10                title  \
0  9780002005883  0002005883               Gilead   
1  9780002261982  0002261987         Spider's Web   
2  9780006178736  0006178731       Rage of angels   
3  9780006280897  0006280897       The Four Loves   
4  9780006280934  0006280935  The Problem of Pain   

                           authors                     categories  \
0               Marilynne Robinson                        Fiction   
1  Charles Osborne;Agatha Christie  Detective and mystery stories   
2                   Sidney Sheldon                        Fiction   
3              Clive Staples Lewis                 Christian life   
4              Clive Staples Lewis                 Christian life   

                                           thumbnail  \
0  http://books.google.com/books/content?id=KQZCP...   
1  http://books.google.com/books/content?id=gA5GP...   
2  http://books.google.com/books/content?id=FKo2T...   
3  http://books.google.com/books/content?id=XhQ5X...   
4  http://books.google.com/books/content?id=Kk-uV...   

                                         description  published_year  \
0  A NOVEL THAT READERS and critics have been eag...          2004.0   
1  A new 'Christie for Christmas' -- a full-lengt...          2000.0   
2  A memorable, mesmerizing heroine Jennifer -- b...          1993.0   
3  Lewis' work on the nature of love divides love...          2002.0   
4  "In The Problem of Pain, C.S. Lewis, one of th...          2002.0   

   average_rating  num_pages  ratings_count     title_and_subtiles  \
0            3.85      247.0          361.0                 Gilead   
1            3.83      241.0         5164.0  Spider's Web: A Novel   
2            3.93      512.0        29532.0         Rage of angels   
3            4.15      170.0        33684.0         The Four Loves   
4            4.09      176.0        37569.0    The Problem of Pain   

                                  tagged_description simple_categories  \
0  9780002005883 A NOVEL THAT READERS and critics...           Fiction   
1  9780002261982 A new 'Christie for Christmas' -...           Fiction   
2  9780006178736 A memorable, mesmerizing heroine...           Fiction   
3  9780006280897 Lewis' work on the nature of lov...        Nonfiction   
4  9780006280934 "In The Problem of Pain, C.S. Le...        Nonfiction   

           isbn  
0           NaN  
1  9.780002e+12  
2           NaN  
3  9.780006e+12  
4  9.780006e+12

Now all the categories are filled in as either **'Fiction'** or **'Non-Fiction'** in the `simple_categories` column.

This simplification helps the model focus on broader, more meaningful groupings, making it easier to **filter**, **classify**, and **recommend** books — without being overwhelmed by hundreds of niche or inconsistent category labels.


In [32]:
# Define the path to save the CSV file in Google Drive
save_path = '/content/drive/MyDrive/Colab Notebooks/Semantic Book Recommender/books_with_simple_categories.csv'

# Save the DataFrame to a CSV file
books.to_csv(save_path, index=False)

print(f"DataFrame successfully saved to {save_path}")

DataFrame successfully saved to /content/drive/MyDrive/Colab Notebooks/Semantic Book Recommender/books_with_simple_categories.csv


In [33]:
filtered_books = books[books["categories"].str.lower().isin([
    "romance",
    "science fiction",
    "scifi",
    "fantasy",
    "horror",
    "mystery",
    "thriller",
    "comedy",
    "crime",
    "historical"
])]

print(f"Number of books in the filtered DataFrame: {len(filtered_books)}")

Number of books in the filtered DataFrame: 16


As we can see, there are only **16 books** in these popular categories. This shows that we **don’t have enough labeled data** to confidently determine their accuracy.

We need to be **extra careful**, as the model might not reliably predict these categories.
